### PR TITLE
Simpler concepts splitter for classification and inputs to concepts attributions

### PR DESCRIPTION
**Pull request by AntoninPoche**
This PR can be reviewed commit by commit. About 5.5k line modifications are tutorial updates.

## Input-to-Concept Attributions

Adds the ability to attribute concept activations back to input tokens using perturbation-based methods. This connects the concept framework with the attribution framework. It is limited to the `*ForSequenceClassification` models.

The idea is that once the concept explainers have been fitted, we can extract their `inputs-to-concepts` and pass it to the attribution methods like a normal model.

`ModelWithSplitPoints` is overcomplex; I made a specific version that forces the split point and granularity to the [CLS] token.

### Key additions

- **`SplitSequenceClassification`**: A `ModelWithSplitPoints` specialization for classification models that auto-detects the classification head as a split point. Provides `inputs_to_activations` and `activations_to_outputs` for efficient concept pipelines.
- **`concept_explainer.inputs_to_concepts`**: Property on all concept explainers returning a `ModelForInputsToConcepts` bridge model (inputs → latent activations → concept activations). Quacks like a `PreTrainedModel` so attribution methods can consume it directly.
- **`InputsToConceptsInferenceWrapper`**: Inference wrapper handling the bridge model inside attribution explainers. Rejects gradient-based methods (incompatible with nnsight tracing).
- **`InputsToConceptsAttributionsExplainer`**: Task-specific attribution explainer auto-selected via `MultitaskExplainerMixin` when the model is a `ModelForInputsToConcepts`.

### Minimal example

```python
import torch
from interpreto import Occlusion, SplitSequenceClassification
from interpreto.concepts import SemiNMFConcepts

# 1. Wrap a classification model
split_model = SplitSequenceClassification(
    "nateraw/bert-base-uncased-emotion", batch_size=32, device_map="cuda"
)

# 2. Fit concepts
activations = split_model.get_activations(train_texts, tqdm_bar=True)
concept_explainer = SemiNMFConcepts(split_model, nb_concepts=20, device="cuda")
concept_explainer.fit(activations)

# 3. Attribute input tokens to concepts
explainer = Occlusion(concept_explainer.inputs_to_concepts, split_model.tokenizer)
results = explainer.explain("The market rallied on strong earnings.", targets=torch.arange(5))
```

### Also includes

- Tests for `SplitSequenceClassification` and input-to-concept attributions
- Documentation: new `interpretations/` docs section with concept attributions page, `SplitSequenceClassification` API page, updated overview

## Type of Change

- 📚 Examples / docs / tutorials / dependencies update
- 🚀 New feature (non-breaking change which adds functionality)

## Checklist

- [x] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/FOR-sight-ai/interpreto/blob/main/CODE_OF_CONDUCT.md) document.
- [x] I've read the [`CONTRIBUTING.md`](https://github.com/FOR-sight-ai/interpreto/blob/main/docs/contributing.md) guide.
- [x] I've successfully run the style checks using `make lint`.
- [x] I've written tests for all new methods and classes that I created and successfully ran `make test`.
- [x] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

File: docs/api/concepts/interpretations/concept_attributions.md
Comment (AntoninPoche) [inline docs/api/concepts/interpretations/concept_attributions.md:1]: Say in the readme and index that we now support this interpretation method.

File: interpreto/model_wrapping/inputs_to_concepts_inference_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    def _extract_targets_from_logits(self, logits):
+        raise NotImplementedError(
+            "InputsToConceptsInferenceWrapper does not support computing targets from logits. "
+            "Concept targets should be provided explicitly."
+        )
+
+    @property
+    def padding_side(self):
+        return "right"
```
Comment (AntoninPoche) [inline interpreto/model_wrapping/inputs_to_concepts_inference_wrapper.py]: Should be an attribute not a property.

File: docs/api/concepts/interpretations/concept_attributions.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+      show_source: true
+      members:
+        - **init**
+        - **call**
+
+::: interpreto.attributions.base.InputsToConceptsAttributionsExplainer
+    handler: python
+    options:
+      show_root_heading: true
+      show_source: true
```
Comment (AntoninPoche) [inline docs/api/concepts/interpretations/concept_attributions.md:122]: Not necessary and generate errors

File: interpreto/model_wrapping/inference_wrapper.py [hunk 3]
```diff
@@ -396,6 +396,8 @@ def _prepare_inputs(self, inputs: list[TensorMapping], for_gradients: bool = Fal
                 padding_side=self.padding_side,
             ).to(self.device)
 
+        padded_inputs.pop("offset_mapping", None)
```
Comment (fanny-jourdan) [inline interpreto/model_wrapping/inference_wrapper.py:399]: What does that mean?

Comment (AntoninPoche) [reply interpreto/model_wrapping/inference_wrapper.py:399]: It removes the `"offset_mapping"` key from the `padded_inputs` dict. But if there is no such key, `None` is returned. We do not need the popped value associated with this key, so the `None` is mainly there to prevent bugs.

In summary, it is a safeguard.

File: interpreto/attributions/base.py [hunk 4]
```diff
@@ -846,13 +851,99 @@ def post_processing(self, contribution: Float[torch.Tensor, "t l"]):
...
+    activation. It bridges the attribution framework with the concept framework:
+    once a concept explainer is fitted, its ``inputs_to_concepts`` property returns
+    a model that can be passed to any perturbation-based attribution method.
+
+    The result is a per-token attribution for each concept, revealing which parts
+    of the input are most responsible for activating a given concept.
+
+    Note:
+        Only perturbation-based methods (Lime, KernelShap, Occlusion, Sobol) are
+        supported. Gradient-based methods are incompatible because the
```
Comment (fanny-jourdan) [inline interpreto/attributions/base.py:867]: Maybe we need to return an error for this ?

File: interpreto/attributions/base.py [hunk 5]
```diff
@@ -846,13 +851,99 @@ def post_processing(self, contribution: Float[torch.Tensor, "t l"]):
         return ModelTask.GENERATION, contribution
 
 
+class InputsToConceptsAttributionsExplainer(AttributionExplainer):
```
Comment (fanny-jourdan) [inline interpreto/attributions/base.py:854]: So smooth ! I like this class :)

File: interpreto/concepts/base.py [hunk 6]
```diff
@@ -63,6 +69,104 @@ def wrapper(self: ConceptEncoderExplainer, *args, **kwargs) -> MethodOutput:
...
+    def __call__(self, **kwargs):
+        """Run inputs → activations → concepts and return a BaseModelOutput-like object.
+
+        Returns:
+            SimpleNamespace with a ``.logits`` attribute containing concept activations.
+        """
+        activations: Float[torch.Tensor, "n d"] = self.split_model.inputs_to_activations(kwargs)
+
+        # TODO: use `encode_activations` which should be renamed `activations_to_concepts`
+        concepts = self.concept_model.encode(activations)
```
Comment (fanny-jourdan) [inline interpreto/concepts/base.py:131]: why you use .encode here ? The name changed no ?

File: interpreto/model_wrapping/split_sequence_classification.py [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+
+        activations = torch.cat(activations, dim=0)
+        predictions = torch.cat(predictions, dim=0)
+
+        # free memory
+        torch.cuda.empty_cache()
+        gc.collect()
+        return {self.classification_head_name: activations, "predictions": predictions}
+
+    def _get_concept_output_gradients(  # type: ignore
```
Comment (fanny-jourdan) [inline interpreto/model_wrapping/split_sequence_classification.py:306]: I'm not sure I understand this function, since you mentioned earlier that the attributions were only made for the perturbation based methods.

File: interpreto/model_wrapping/split_sequence_classification.py [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+        **kwargs,  # not used, just to support the `model_with_split_points` interface
+    ) -> LatentActivations:
+        """Extract activations for the specified split point."""
+        return activations[self.classification_head_name]
+
+    def get_latent_shape(  # type: ignore
+        self,
+        **kwargs,  # not used, just to support the `model_with_split_points` interface
+    ) -> dict[str, torch.Size]:
+        """Get the shape of the latent activations at the specified split point.
```
Comment (fanny-jourdan) [inline interpreto/model_wrapping/split_sequence_classification.py]: why you write "specified split point" since we dont have split point here (only splitted before classification head) ?

File: interpreto/model_wrapping/inputs_to_concepts_inference_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+import torch
+from beartype import beartype
+from jaxtyping import Float, Int, jaxtyped
+
+from interpreto.model_wrapping.inference_wrapper import InferenceWrapper
+from interpreto.typing import IncompatibilityError
+
+
+class InputsToConceptsInferenceWrapper(InferenceWrapper):
```
Comment (fanny-jourdan) [inline interpreto/model_wrapping/inputs_to_concepts_inference_wrapper.py:35]: I don't really understand the inference wrapper or why it's “so simple”. Shouldn't there be something that goes from the input to the CLS token, and then from the CLS token to the concept space?

File: interpreto/attributions/base.py (hunk 4 above)
Comment (AntoninPoche) [reply interpreto/attributions/base.py:867]: It does return an error, in the `InputsToConceptsInferenceWrapper.__init__()`. It clearly states that it is incompatible and only works for perturbation-based.

File: interpreto/concepts/base.py (hunk 6 above)
Comment (AntoninPoche) [reply interpreto/concepts/base.py:131]: The name will be changed in #156. I deemed it a refactoring because it affects more files. And I could not update only half of them.

File: interpreto/model_wrapping/split_sequence_classification.py (hunk 8 above)
Comment (AntoninPoche) [reply interpreto/model_wrapping/split_sequence_classification.py]: It is an artifact and a mistake; I will remove it when taking your comments into account.

File: interpreto/model_wrapping/inputs_to_concepts_inference_wrapper.py (hunk 9 above)
Comment (AntoninPoche) [reply interpreto/model_wrapping/inputs_to_concepts_inference_wrapper.py:35]: Most of the magic is done in `interpreto.concepts.base.ModelForInputsToConcepts`, this model is constructed from the concept explainer, and its `__call__` function corresponds to inputs to concepts. It does `inputs_to_activations` (from the input to the [CLS] token), then `activations_to_concepts` (from the [CLS] token to the concepts).

In other words, the inference wrapper tells you that the model is an `InputsToConcepts` and thus inputs are the same and outputs are the concepts.

The reason it is "so simple" is because the `attributions` module was nicely designed ^^, but also because it takes text as input and concepts are not that far from class logits (once you do not care how you go from one to the other).

Comment (fanny-jourdan) [reply interpreto/model_wrapping/inputs_to_concepts_inference_wrapper.py:35]: Ok it s clearer now thanks !

File: tests/concepts/interpretation/test_inputs_to_concepts_attributions.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    Test using the input-to-concepts model in an attribution explainer.
+
+    This test is slow because it tests with multiple models and explainers.
+
+    Not all combinations are tested, there are far too many, but we cycle over the REPRESENTATIVE_CASES.
+    """
+    inputs_to_concepts_attributions(repo_id, concepts_explainer_class, attribution_explainer_class, sentences)
+
+
+def inputs_to_concepts_attributions(
```
Comment (fanny-jourdan) [inline tests/concepts/interpretation/test_inputs_to_concepts_attributions.py:117]: why the function just named "inputs_to_concepts_attributions" ? It s a test function, it is not better to name it with "test" something ? (it s a detail but for me it would be clearer)

**Review by fanny-jourdan — APPROVED** [review]
Good work, Antonin, and thanks for responding to my comments as I went along, it’s clearer now! I just added a quick note in one of the tests, but other than that, everything looks good to me!

File: tests/concepts/interpretation/test_inputs_to_concepts_attributions.py (hunk 10 above)
Comment (AntoninPoche) [reply tests/concepts/interpretation/test_inputs_to_concepts_attributions.py:117]: With `pytest`, every function starting with `test_` is executed; the rest are ignored.

The reason this function does not start with `test_` is that other functions call it, in particular:

```python
# Only runs the test once
def test_inputs_to_concepts_attributions_fast(sentences):
    inputs_to_concepts_attributions("hf-internal-testing/tiny-random-distilbert", PCAConcepts, KernelShap, sentences)

# Runs the test for all representative cases
@pytest.mark.slow
@pytest.mark.parametrize("repo_id, concepts_explainer_class, attribution_explainer_class", REPRESENTATIVE_CASES)
def test_inputs_to_concepts_attributions_slow(
    repo_id, concepts_explainer_class, attribution_explainer_class, sentences
):
    inputs_to_concepts_attributions(repo_id, concepts_explainer_class, attribution_explainer_class, sentences)
```

Note that the second one has `@pytest.mark.slow`. Which is ignored in `make fast-test` (the command used in our GitHub workflow and pre-commit). This way, we can quickly test for big regression with `make fast-test`. But also ensure everything works fine with `make test`. This duality is useful depending on what you modify in the library. The full test does not need to be run if you do not touch anything that could affect it. Still, from time to time, full tests are useful.

For few other tips, the commands behind are the following:
`make test`: `python -m pytest -n auto -c pyproject.toml -v`
`make fast-test`: `python -m pytest -n auto -c pyproject.toml -v -m "not slow"`

But you can also test only a subset of elements (to go faster):
`python -m pytest -n auto -c pyproject.toml -v -m "not slow" tests/attributions/methods/`
